### PR TITLE
feat: config options for fuzzy search

### PR DESCRIPTION
**Pull request by johannes-vogel**

### CODE REVIEW

File: hana/lib/cql-functions.js [hunk 1]
```diff
@@ -24,10 +24,39 @@ const StandardFunctions = {
...
+    let fuzzyString
+
+    // default config
+    const fuzzyIndex = cds.env.hana?.fuzzy || 0.7
+
+    // if column specific value is provided, the configuration has to be defined on column level
+    if (csnElements.some(e => e.element?.['@Search.ranking'] || e.element?.['@Search.fuzzinessThreshold'])) {
+      const cols = csnElements.map(e => {
+        // REVISIT: How to do quoting?
+        let col = `${e.ref.join('.')} FUZZY`
```
Comment (patricebender) [inline hana/lib/cql-functions.js:37]: there is a `quote` function in `cqn2sql` which takes care of quotation, if necessary. Bind the `this` of the `cqn2sql` instance and use it here, I would say.

Comment (johannes-vogel) [reply hana/lib/cql-functions.js:37]: Bob so far uses a different technique by changing the toString method. That does not suite my use case but I don't know the implications of changing it. I propose to discuss later

File: hana/lib/cql-functions.js [hunk 2]
```diff
@@ -24,10 +24,68 @@ const StandardFunctions = {
...
+    // if column specific value is provided, the configuration has to be defined on column level
+    if (csnElements.some(e => e.element?.['@Search.ranking'] || e.element?.['@Search.fuzzinessThreshold'])) {
+      csnElements.forEach(e => {
+        let fuzzy = `FUZZY`
+        
+        // weighted search
+        const rank = e.element?.['@Search.ranking']?.['=']
+        if(rank === 'HIGH') fuzzy += ' WEIGHT 0.8'
+        else if(rank === 'LOW') fuzzy += ' WEIGHT 0.3'
+        else fuzzy += ' WEIGHT 0.5' // MEDIUM
```
Comment (BobdenOs) [inline hana/lib/cql-functions.js:65]: Doesn't java `throw` when an invalid variable is defined ? I am fine either way.

File: hana/lib/cql-functions.js [hunk 3]
```diff
@@ -24,10 +24,68 @@ const StandardFunctions = {
...
+        
+        // fuzziness
+        fuzzy+= ` MINIMAL TOKEN SCORE ${e.element?.['@Search.fuzzinessThreshold'] || fuzzyIndex}`
+        fuzzy+= " SIMILARITY CALCULATION MODE 'search'"
```
Comment (BobdenOs) [inline hana/lib/cql-functions.js:69]: Could be a single template string 🤷

File: hana/test/fuzzy.test.js [hunk 4]
```diff
@@ -1,19 +1,78 @@
...
+    test('annotations', async () => {
+      const { BooksAnnotated } = cds.entities('sap.capire.bookshop')
+      const cqn = SELECT.from(BooksAnnotated).search('"heights"').columns('1')
+      const {sql} = cqn.toSQL()
+      expect(sql).to.include('title FUZZY WEIGHT 0.8 MINIMAL TOKEN SCORE 0.9')
+      expect(sql).to.include('code FUZZY WEIGHT 0.5 MINIMAL TOKEN SCORE 0.7')
+      expect(sql).to.include('descr FUZZY WEIGHT 0.3 MINIMAL TOKEN SCORE 0.6')
+      
+      const res = await SELECT.from(BooksAnnotated).search('"heights"')
+      expect(res[0].title).to.eq('Wuthering Heights')
```
Comment (BobdenOs) [inline hana/test/fuzzy.test.js:50]:
```suggestion
      const { title }= await SELECT.one.from(BooksAnnotated).search('"heights"')
      expect(title).to.eq('Wuthering Heights')
```

File: hana/lib/cql-functions.js [hunk 5]
```diff
@@ -24,10 +24,63 @@ const StandardFunctions = {
...
```
Comment (johannes-vogel) [inline hana/lib/cql-functions.js:69]:
```suggestion
        fuzzy+= ` MINIMAL TOKEN SCORE ${e.element?.['@Search.fuzzinessThreshold'] || fuzzyIndex} SIMILARITY CALCULATION MODE 'search'`
```

File: test/cds.js [hunk 6]
```diff
@@ -72,7 +72,7 @@ cds.test = Object.setPrototypeOf(function () {
         hash.update(isolateName)
         ret.data.isolation = isolate = {
           // Create one database for each overall test execution
-          database: process.env.TRAVIS_JOB_ID || process.env.GITHUB_RUN_ID || require('os').userInfo().username || 'test_db',
+          database: process.env.TRAVIS_JOB_ID || process.env.GITHUB_RUN_ID || require('os').userInfo().username.repeat(6) || 'test_db',
```
Comment (johannes-vogel) [inline test/cds.js:75]:
```suggestion
          database: process.env.TRAVIS_JOB_ID || process.env.GITHUB_RUN_ID || require('os').userInfo().username || 'test_db',
```

**Pull request merged**